### ISSUE TITLE
Error connecting to the receptor.

### ISSUE THREAD

**Issue by smugcloud** [cloudfoundry-incubator/lattice-release#215]
I am following the [tutorial instructions] (http://lattice.cf/docs/getting-started/) on OSX 10.10.5 with VirtualBox.  I can confirm the target is set, but when I try to create an app, I keep getting `Error connecting to the receptor.`.  This is similar to #207, but executing the commands mentioned there get me back to the same place.  I have tried a personal docker image that is available on the machine, as well as the sample `cloudfoundry/lattice-app`.  Am I missing something or is this a known issue?

```
$ ./ltc --version
ltc version v0.5.0 (diego 0.1424.1)

$ ./ltc target
Target:		192.168.11.11.xip.io
Droplet store:	192.168.11.11.xip.io:8444

$ ./ltc create lattice-app cloudfoundry/lattice-app
Error connecting to the receptor. Make sure your lattice target is set, and that lattice is up and running.
	Underlying error: Get http://receptor.192.168.11.11.xip.io/v1/desired_lrps: dial tcp 192.168.11.11:80: i/o timeout

$ ./ltc debug-logs
Error connecting to the receptor. Make sure your lattice target is set, and that lattice is up and running.
	Underlying error: Get http://receptor.192.168.11.11.xip.io/v1/desired_lrps: dial tcp 192.168.11.11:80: i/o timeout

$ curl -v http://receptor.192.168.11.11.xip.io/v1/desired_lrps
*   Trying 192.168.11.11...
* Connected to receptor.192.168.11.11.xip.io (127.0.0.1) port 80 (#0)
> GET /v1/desired_lrps HTTP/1.1
> Host: receptor.192.168.11.11.xip.io
> User-Agent: curl/7.43.0
> Accept: */*
>
```

**Comment by davidwadden** [comment]
That error doesn't mean much unfortunately, the API endpoint (receptor) decided not to answer within 5 seconds, so it gives the i/o timeout.  Can you try this:

```bash
$ LATTICE_SYSTEM_DOMAIN=local.lattice.cf vagrant up
$ ltc target local.lattice.cf
```

I want to exclude any weirdness with `xip.io` DNS resolution.  The target being set doesn't mean much, a better test is to run `ltc test -v` after bringing up the cluster, and then testing connectivity with `ltc ls`.  What's that do?

**Comment by amaltson** [comment]
@davidwadden I've found something very strange... I'm getting this same `dial tcp: i/o timeout` issue with the receptor on our custom domain. But here's the kicker, it only seems to happen to Mac machines and only after the first day. I had a coworker try Lattice on his Mac yesterday, and all `ltc` commands were coming through perfectly. I asked him today to run `ltc ls` or any of the other commands and he started experiencing the same `dial tcp: i/o timeout` issues.

```
$ ltc ls
Error connecting to the receptor. Make sure your lattice target is set, and that lattice is up and running.
	Underlying error: Get http://receptor.<our custom domain>/v1/desired_lrps: dial tcp: i/o timeout
```

I've had this happen to two people so far, but a Linux machine I ssh into continues to work fine every day. They're all on the same subnet networking wise

It seems to be an issue with the Mac's networking, because I've even spun up a Vagrant Ubuntu machine (NAT or bridge), put the Lattice CLI on it and found the same timeout issue.

This only seems to affect Mac machines, and while we run Lattice 0.4.3 on the backend, I've tried with every client from 0.4.3 to 0.5.0.

I'll try to do some more `tcpdump` analysis tomorrow if I can dig out more info.

**Comment by davidwadden** [comment]
We see all sorts of flakiness around OS X DNS resolution that I think align well with your report.

Did you try updating `/etc/resolv.conf` on the linux box to **not** resolve through the host (setting to, say, `8.8.8.8`)?  I've seen that help sometimes.

Otherwise, to the root cause, have you tried following our [troubleshooting](http://lattice.cf/docs/troubleshooting/#ltc-is-giving-no-such-host-errors-with-xip-io-addresses-help) page for these issues?  We typically have luck by restarting the DNS service on OS X.  `discoveryd` in particular has proved problematic.

**Comment by amaltson** [comment]
Hey @davidwadden, I tried changing the resolution to go to `8.8.8.8`, but that doesn't work because the Lattice server is all private. When I only point at our nameserver (not the private Vagrant network) it times out.

I don't think this is a DNS issue though, because from the Mac itself the DNS resolution seems very fast:

```
$ time host cf.example.com
cf.example.com has address 192.168.0.20
host cf.example.com  0.00s user 0.00s system 74% cpu 0.008 total
```

Also, when I `tcpdump` the IP address of the Lattice server and telnet to it, I get a bunch of output:

```
$ telnet cf.example.com 80
Trying 192.168.0.20...
Connected to cf.example.com.
Escape character is '^]'.
$ sudo tcpdump -i en5 -n -s0 -ttt host 192.168.0.20
tcpdump: verbose output suppressed, use -v or -vv for full protocol decode
listening on en5, link-type EN10MB (Ethernet), capture size 65535 bytes
00:00:00.000000 ARP, Request who-has 192.168.0.20 tell 192.168.0.153, length 28
00:00:00.000426 ARP, Reply 192.168.0.20 is-at 00:0c:29:66:32:49, length 46
00:00:00.000013 IP 192.168.0.153.62621 > 192.168.0.20.80: Flags [S], seq 3808710102, win 65535, options [mss 1460,nop,wscale 5,nop,nop,TS val 628759489 ecr 0,sackOK,eol], length 0
00:00:00.000319 IP 192.168.0.20.80 > 192.168.0.153.62621: Flags [S.], seq 971154430, ack 3808710103, win 28960, options [mss 1460,sackOK,TS val 80913569 ecr 628759489,nop,wscale 7], length 0
00:00:00.000028 IP 192.168.0.153.62621 > 192.168.0.20.80: Flags [.], ack 1, win 4117, options [nop,nop,TS val 628759489 ecr 80913569], length 0
00:00:05.000049 ARP, Request who-has 192.168.0.153 tell 192.168.0.20, length 46
00:00:00.000032 ARP, Reply 192.168.0.153 is-at 40:6c:8f:51:63:0d, length 28
```

But when I do a `ltc ls` or any other Lattice command the `tcpdump` is silent.

The other thing I've noticed, locally when I initially curl the receptor API, it takes more than 5 seconds to respond. This makes me thing of another comment you made [on another issue](https://github.com/cloudfoundry-incubator/lattice-release/issues/128#issuecomment-112225531) that the Go client has a timeout of 5 seconds and if the receptor doesn't respond in that time it times out.

```
$ time curl http://receptor.cf.example.com/v1/desired_lrps
[{"process_guid":"hossein","domain":"lattice","rootfs":"preloaded:cflinuxfs2","instances":1,"env":[{"name":"PROCESS_GUID","value":"hossein"},{"name":"MEMORY_LIMIT","value":"128M"}
...
0.00s user 0.01s system 0% cpu 5.590 total
$ time curl http://receptor.cf.example.com/v1/desired_lrps
...
 0.00s user 0.00s system 0% cpu 2.571 total
$ time curl http://receptor.cf.example.com/v1/desired_lrps
...
 0.00s user 0.00s system 16% cpu 0.042 total
```

As you can see, subsequent calls seem to cache the result and happen much faster.

I'm in the process of setting up Lattice on another machine equipped with SSDs, my hope is that solves the problem. Thanks again!

**Comment by davidwadden** [comment]
Missed this one..  This is a lot to parse.

Let me get a better idea of your setup.

1.  Single VM Lattice (colocated) running on VirtualBox
  - Host-only network defined on 192.168.0.0/24
  - Lattice IP=192.168.0.20;  Host IP=192.168.0.153
  - DNS:  served elsewhere on network  wildcard DNS record on cf.example.com resolves to Lattice IP (192.168.0.20)
  - Was this started using `LATTICE_IP=192.168.0.20 vagrant up` ?  Or is there custom network configuration in the `Vagrantfile` ?
2. Do you have any custom resolver config in OS X?
  - We had a lot of issues trying to recommend using dnsmasq and don't recommend custom resolvers.
3. Do you run with any proxies?
  - `env | grep -i proxy`  

The i/o error totally makes sense is curl is taking 6 seconds.  There's some network latency, not sure if it's DNS or otherwise.  So trying to find out what the transport layer is for your `ltc` communication, it's entirely across a VirtualBox virtual network?  Or is it bridged out to your LAN there?

Any difference with Lattice v0.6.0 ?

**Comment by amaltson** [comment]
@davidwadden sorry, let me answer these questions.

1. Single VM Lattice (colocated) running on **vSphere**.
 - Shared network
 - Those IPs were examples, but yes.
 - Yes, the DNS resolves to the correct IP.
 - `LATTICE_IP` is in `/var/lattice/setup/lattice-environment`
2. No custom resolver configured in OS X
3. No proxy, `$ env | grep proxy` returns nothing.

I just tried deploying a Lattice 0.5.0 instance and now running into another issue. This time I am running with the proxy set as per the [documentation](http://lattice.cf/docs/proxy/)

```
$ ltc bd badges ruby
Uploading application bits...
Uploaded.
Submitted build of badges
11/06 14:34:12.07 [BUILD|0] Error deleting http://cf.example.com:8444/blobs/badges/bits.zip: dial tcp: i/o timeout
11/06 14:34:12.07 [BUILD|0] Exit status 2
Build failed: Exited with status 2
```

I can't use Lattice v0.6.0 because there's no vSphere support that I could see with the `packer-bosh`.

**Comment by amaltson** [comment]
@davidwadden I was reading through [Julia Evan's excellent blog post](http://jvns.ca/blog/2015/11/21/why-you-should-understand-a-little-about-tcp/) about understanding a bit about TCP, and it hit me. Could this discrepancy between the Lattice CLI on Linux vs OS X have anything to do with `TCP_NODELAY`? If it slows down the connection long enough that it passes the 5 second limit on the Lattice CLI? Is this possible?

**Comment by sclevine** [comment]
Lattice was a Cloud Foundry project focused on delivering the Diego application scheduler into the hands of users as quickly as possible before Diego was fully integrated to Cloud Foundry. Lattice served that purpose well, but now Diego is available as the application scheduler for Cloud Foundry, and Cloud Foundry is the best way to use Diego. Therefore, Lattice is no longer actively maintained.

Apologies for not continuing to address your issue.

-Stephen
